File: src/main/java/ru/nucodelabs/algorithms/charts/MisfitValuesFactoryNative.java
package ru.nucodelabs.algorithms.charts;

import ru.nucodelabs.algorithms.forward_solver.ForwardSolver;
import ru.nucodelabs.data.ves.ExperimentalMeasurement;
import ru.nucodelabs.data.ves.ModelLayer;

import java.util.ArrayList;
import java.util.List;

import static java.lang.Math.abs;
import static java.lang.Math.signum;

final class MisfitValuesFactoryNative implements MisfitValuesFactory {

    @Override
    public List<Double> apply(List<ExperimentalMeasurement> experimentalData, List<ModelLayer> modelData) {
        List<Double> resistanceApparent = experimentalData.stream().map(ExperimentalMeasurement::getResistanceApparent).toList();
        List<Double> errorResistanceApparent = experimentalData.stream().map(ExperimentalMeasurement::geErrorResistanceApparent).toList();

        if (experimentalData.size() == 0 || modelData.size() == 0) {
            return new ArrayList<>();
        }

        ForwardSolver forwardSolver = ForwardSolver.createDefaultForwardSolver(experimentalData, modelData);

        List<Double> solvedResistance = forwardSolver.solve();

        List<Double> res = new ArrayList<>();

        for (int i = 0; i < experimentalData.size(); i++) {
            double value = abs(MisfitFunctions.calculateRelativeDeviationWithError(
                    resistanceApparent.get(i),
                    errorResistanceApparent.get(i) / 100f,
                    solvedResistance.get(i)
            )) * signum(solvedResistance.get(i) - resistanceApparent.get(i)) * 100f;
            res.add(value);
        }

        return res;
    }
}

File: src/main/java/ru/nucodelabs/algorithms/charts/MisfitValuesFactoryNative.kt
package ru.nucodelabs.algorithms.charts

import ru.nucodelabs.algorithms.forward_solver.ForwardSolver
import ru.nucodelabs.data.ves.ExperimentalData
import ru.nucodelabs.data.ves.ModelLayer
import kotlin.math.abs
import kotlin.math.sign

internal class MisfitValuesFactoryNative(val forwardSolver: ForwardSolver) : MisfitsFunction {

    override fun invoke(experimentalData: List<ExperimentalData>, modelData: List<ModelLayer>): List<Double> {
        if (experimentalData.isEmpty() || modelData.isEmpty()) {
            return listOf()
        }

        val resistanceApparent = experimentalData.map { it.resistanceApparent }
        val errorResistanceApparent = experimentalData.map { it.errorResistanceApparent }

        val solvedResistance = forwardSolver(experimentalData, modelData)
        val res = mutableListOf<Double>()
        for (i in experimentalData.indices) {
            val value = abs(
                MisfitFunctions.calculateRelativeDeviationWithError(
                    resistanceApparent[i],
                    errorResistanceApparent[i] / 100f,
                    solvedResistance[i]
                )
            ) * sign(solvedResistance[i] - resistanceApparent[i]) * 100f

            res.add(value)
        }
        return res
    }
}